### ISSUE TITLE
is registered as 'Execution Context Scope' lifestyle, but the instance is requested outside the context of a Execution Context Scope.

### ISSUE THREAD

**Issue by stevengopmo** [simpleinjector/SimpleInjector#334]
I'm getting the following error from my container

> The ClaimsPrincipal is registered as 'Execution Context Scope' lifestyle, but the instance is requested outside the context of a Execution Context Scope.

I'm assigning a listener to my RavenDB DocumentStore (singleton) and the listener constructor has the following constructor signature.

`public AuditableEntityListener(Func<Collaborator> getCollaborator, IClock clock);`

Since I'm using the `Func<Collaborator>`, I thought this would get injected as a singleton also (per a previous discussion here #253).  IClock is also singleton.  

Here is the complete registration
`            store.RegisterListener(new AuditableEntityListener(() =>
            {
                var principal = container.GetInstance<ClaimsPrincipal>();
                return principal.AsCollaborator();
            }, SystemClock.Instance));`

Also, it doesn't seem I always get this error, but that could be due to current usage.

In case it makes any difference I'm using version 3.1.5.

**Comment by dotnetjunkie** [comment]
The #253 discussion talks especially about `Func<T>` decoratee dependencies. Simple Injector understands `Func<T>` dependencies inside decorators.

In all other cases, Simple Injector does not automatically inject `Func<T>` dependencies for you. This is a delibarate design decision, and is stated [here](https://simpleinjector.readthedocs.io/en/latest/howto.html#register-factory-delegates):

> Note: We personally think that allowing to register `Func<T>` delegates by default is a design smell. The use of `Func<T>` delegates makes your design harder to follow and your system harder to maintain and test. Your system should only have a few of those factories at most. If you have many constructors in your system that depend on a `Func<T>`, please take a good look at your dependency strategy. [The following article](https://www.cuttingedge.it/blogs/steven/pivot/entry.php?id=100) goes into details about why Abstract Factories (such as `Func<T>`) are a design smell.

If the use of the `Func<Collaborator>` is an absolute must, you can add the following registration:

``` c#
container.RegisterSingleton<Func<Collaborator>>(container.GetInstance<Collaborator>);
```

**Comment by stevengopmo** [comment]
I agree with the premise in general.  In this particular case, I am trying to wire context (execution scope) sensitive object instance to a singleton object.  If there's a clearer way of doing it, I'm all ears! :)

**Comment by dotnetjunkie** [comment]
Since your document store is a singleton, and you need to access a scoped object, you will have to resolve it on the spot. So what you're doing here seems sane to me.

Why you are getting this "the instance is requested outside the context of a Execution Context Scope" exception is unclear to me; you haven't provided any details (such as a stack trace). There are some typical causes why this can happen though:

 - You didn't wrap the operation in a `container.BeginExecutionContextScope()` call.
 - Somewhere during the asynchronous flow of methods, `ConfigureAwait(false)` is called. `ConfigureAwait(false)` causes the CLR to ditch its contextual information. The `ExecutionContextScope` requires this context to be available.
 - The operation executed at a point where the framework already quit the asynchronous scope. A [notorious framework](https://stackoverflow.com/a/39974586/264697) that does this is ASP.NET Web Api, during disposes the `IDependencyScope`.

**Comment by stevengopmo** [comment]
as to wrapping the operation in a `container.BeginExecutionContextScope()` I'm actually doing this in the same project as discussed in #253 so while it's possible I've got something screwy, I think I'm doing it right.

here's the stack trace.
`11/21/2016 8:42:49 PM -05:00: SimpleInjector.ActivationException: The registered delegate for type ClaimsPrincipal threw an exception. The ClaimsPrincipal is registered as 'Execution Context Scope' lifestyle, but the instance is requested outside the context of a Execution Context Scope. ---> SimpleInjector.ActivationException: The ClaimsPrincipal is registered as 'Execution Context Scope' lifestyle, but the instance is requested outside the context of a Execution Context Scope.
  at SimpleInjector.Scope.GetScopelessInstance[TService,TImplementation](ScopedRegistration`2 registration)
  at SimpleInjector.Scope.GetInstance[TService,TImplementation](ScopedRegistration`2 registration, Scope scope)
  at SimpleInjector.Advanced.Internal.LazyScopedRegistration`2.GetInstance(Scope scope)
  at DynamicInstanceProducer2.GetInstance(Object[] constants)
  at SimpleInjector.Internals.CompilationHelpers.ConstantsClosure`1.GetInstance()
  at SimpleInjector.InstanceProducer.GetInstance()
  --- End of inner exception stack trace ---
  at SimpleInjector.InstanceProducer.GetInstance()
  at SimpleInjector.Container.GetInstance[TService]()
  at goPMO.Impel.Service.CompositionRoot.Bootstrapper.<>c__DisplayClass0_0.<Bootstrap>b__5() in C:\projects\impel-api-2-0\source\goPMO.Impel.Service\CompositionRoot\Bootstrapper.cs:line 208
  at goPMO.Common.RavenDB.AuditableEntityListener.BeforeStore(String key, Object entityInstance, RavenJObject metadata, RavenJObject original)
  at Raven.Client.Document.InMemoryDocumentSessionOperations.PrepareForEntitiesPuts(SaveChangesData result) in c:\Builds\RavenDB-Stable-3.0\Raven.Client.Lightweight\Document\InMemoryDocumentSessionOperations.cs:line 1004
  at Raven.Client.Document.InMemoryDocumentSessionOperations.PrepareForSaveChanges() in c:\Builds\RavenDB-Stable-3.0\Raven.Client.Lightweight\Document\InMemoryDocumentSessionOperations.cs:line 984
  at Raven.Client.Document.Async.AsyncDocumentSession.<SaveChangesAsync>d__d1.MoveNext() in c:\Builds\RavenDB-Stable-3.0\Raven.Client.Lightweight\Document\Async\AsyncDocumentSession.cs:line 917
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  at Rebus.RavenDb.Timouts.RavenDbTimeoutManager.<Defer>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  at Rebus.Pipeline.Receive.HandleDeferredMessagesStep.<StoreMessageUntilDue>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  at Rebus.Pipeline.Receive.HandleDeferredMessagesStep.<Process>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  at Rebus.Retry.Simple.SimpleRetryStrategyStep.<DispatchWithTrackerIdentifier>d__8.MoveNext()
`

**Comment by dotnetjunkie** [comment]
Can you show what's happening around line 208 of your `goPMO.Impel.Service\CompositionRoot\Bootstrapper.cs` class?

**Comment by stevengopmo** [comment]
line 208 is the `container.GetInstance<ClaimsPrincipal>(); ` from the registration of AuditableEntityListener

**Comment by dotnetjunkie** [comment]
If you look through the call stack, the `LifetimeScopeDecorator` is nowhere to be found. I suspect that the `RavenDbTimeoutManager` gets triggered by some windows `Timer` on a background thread, while you don't wrap this operation in an execution context scope.

**Comment by stevengopmo** [comment]
ugh.  You are right I think.  Drat, so what's the answer?  Register the Func<> as singleton?  It seems that would be correct as the Rebus bus is singleton as well and the timeout manager is dependent on store (singleton).

**Comment by dotnetjunkie** [comment]
It should be no problem that the bus and timeout manager are singletons, as long as they allow you to intercept somewhere down the pipeline and start a request (by using `BeginExecutionContextScope()`).

**Comment by stevengopmo** [comment]
sorry @dotnetjunkie that went over my head. :)   I **THINK** you are saying this is ok as long as there is someway to capture when a timeout manager is created and start the execution scope.

The bus is a singleton but I don't know that the timeout manager is, only that it depends on the store singleton, which has this Func to get the current collaborator based on the context of the message.  Which is where the problem is occurring as you point out.  So while I think registering the Func as a singleton will solve the problem I don't think that is what you are saying.

**Comment by stevengopmo** [comment]
FWIW, rebus does have an external manager option for timeouts that I "should" be using but now I'm determined to fix (AND understand) this for future cases of similar situations.

**Comment by dotnetjunkie** [comment]
Hmm, yes, I'm sorry. That might have been a bit unclear. Let me try again.

To be able to resolve objects that are defined as `Scoped`, an active scope must exist. In the common case, scopes are defined as 'ambient' in Simple Injector. This means that they are always around; they don't have to be passed along the object graph. You just begin a scope and Simple Injector will manage the rest, and it will warn you in case you forgot to start a scope.

When that timer goes off, you need to execute something interesting logic. To be able to do so, you need to resolve an instance. Since that instance is scoped, there must be an active scope. Although you could make this scoped object hybrid and fall back to it being transient in case there is no active scope, this practice is not advised, because it is error phrone. In case the scoped object is a dependency in a graph that you resolve, this model will break once that object gets inject multiple times. In that case you usually want to preserve scoped behavior. In other words: Only have one instance during that scope/graph. Not only that, you typically need deterministic disposal.

For most applications, the Simple Injector integration packages will ensure that scopes are created implicitly for you. This holds for WCF, MVC and Web API. The scope is usually wrapped around a single web request; when the request ends, the scope will be disposed.

Sometimes however, you will have to explicitly create a new scope. For instance:
 - You might want to have an isolated unit of work within the web request, where the request might get a different `DbContext` than the unit of work gets.
 - You might want to run business logic on a background thread, where the integration packages can't automatically create a scope for you.
 - You might have some scheduled task on a background thread that gets triggered using some timer.

This last point is the situation you are in. There is a timer that goes off and you need to wrap it in a scope. Within that scope you need to resolve an instance and run your business logic.

I have no experience with Rebus, but it should be possible to hook into that framework at the level of `AsyncDocumentSession` or `InMemoryDocumentSessionOperations` to make sure that a scope is started. Alternatively, you might be able to replace your `ClaimsPrincipal` with a proxy that begins/ends a scope inside its method call, but it's hard to say from here whether that is possible or not. There are usually several options.

**Comment by stevengopmo** [comment]
Thank you @dotnetjunkie your insight is helpful as always.  I will look for that hook and update this issue when found.

That being said, it seems to me the singleton registration for the Func<ClaimsPrincipal> still has a place in this situation where my singleton store needs a context specific scoped dependency.  Would you agree?

**Comment by dotnetjunkie** [comment]
Yes, I agree. There's nothing wrong with that code.
